File: java-debugger/src/main/java/com/example/microbreakpoint/service/AgentBreakpointService.java
package com.example.microbreakpoint.service;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

@Service
public class AgentBreakpointService {

    private final DebugService debugService;
    private final JdbcTemplate jdbc;

    public AgentBreakpointService(DebugService debugService) {
        this.debugService = debugService;
        this.jdbc = debugService.jdbc();
    }

    public Map<String, Object> declare(Map<String, Object> payload) {
        return debugService.declareBreakpointRule(payload);
    }

    public Map<String, Object> list(String sessionId) {
        List<Map<String, Object>> rules = debugService.listBreakpoints(sessionId).stream()
                .map(this::agentRule)
                .toList();
        return Map.of(
                "ok", true,
                "breakpoint_rules", rules,
                "entities", rules.stream().map(this::entity).toList());
    }

    public Map<String, Object> get(String ruleId) {
        Map<String, Object> row = row(ruleId);
        if (row == null) {
            return error("not_found", ruleId, "断点规则不存在。");
        }
        Map<String, Object> rule = agentRule(row);
        return response(rule, "断点规则已读取。");
    }

    public Map<String, Object> setEnabled(String ruleId, boolean enabled) {
        if (row(ruleId) == null) {
            return error("not_found", ruleId, "断点规则不存在。");
        }
        debugService.setBreakpointEnabled(ruleId, enabled);
        Map<String, Object> rule = agentRule(row(ruleId));
        return response(rule, enabled ? "断点规则已启用。" : "断点规则已禁用。");
    }

    public Map<String, Object> delete(String ruleId) {
        Map<String, Object> row = row(ruleId);
        if (row == null) {
            return error("not_found", ruleId, "断点规则不存在。");
        }
        Map<String, Object> rule = agentRule(row);
        debugService.deleteBreakpoint(ruleId);
        rule.put("status", "cancelled");
        return response(rule, "断点规则已取消。");
    }

    private Map<String, Object> row(String ruleId) {
        List<Map<String, Object>> rows = jdbc.queryForList("SELECT * FROM breakpoint WHERE id=?", ruleId);
        return rows.isEmpty() ? null : debugService.normalize(rows.get(0));
    }

    private Map<String, Object> agentRule(Map<String, Object> row) {
        String objectName = str(firstNonNull(row.get("objectName"), row.get("object_name")));
        String cmdName = str(firstNonNull(row.get("cmdName"), row.get("cmd_name")));
        String displayName = str(firstNonNull(row.get("displayName"), row.get("display_name")));
        String matchMode = str(firstNonNull(row.get("matchMode"), row.get("match_mode")));
        boolean enabled = bool(row.get("enabled"));
        Map<String, Object> target = new LinkedHashMap<>();
        target.put("object", objectName);
        target.put("command", cmdName);
        target.put("display_name", displayName);
        target.put("session_id", firstNonNull(row.get("sessionId"), row.get("session_id")));

        Map<String, Object> match = new LinkedHashMap<>();
        match.put("type", "params_condition".equals(matchMode) ? "parameters" : "interface");
        match.put("mode", matchMode);
        if ("params_condition".equals(matchMode)) {
            match.put("conditions", firstNonNull(row.get("conditions"), List.of()));
        }

        Map<String, Object> rule = new LinkedHashMap<>();
        rule.put("breakpoint_rule_id", row.get("id"));
        rule.put("label", label(objectName, cmdName, displayName));
        rule.put("status", enabled ? "armed" : "disabled");
        rule.put("target", target);
        rule.put("match", match);
        rule.put("hit_count", firstNonNull(row.get("hitCount"), row.get("hit_count"), 0));
        rule.put("source_type", firstNonNull(row.get("sourceType"), row.get("source_type")));
        rule.put("created_at", firstNonNull(row.get("createdAt"), row.get("created_at")));
        rule.put("updated_at", firstNonNull(row.get("updatedAt"), row.get("updated_at")));
        return rule;
    }

    private Map<String, Object> response(Map<String, Object> rule, String message) {
        Map<String, Object> result = new LinkedHashMap<>(rule);
        result.put("ok", true);
        result.put("message", message);
        result.put("entities", List.of(entity(rule)));
        return result;
    }

    private Map<String, Object> entity(Map<String, Object> rule) {
        Map<String, Object> entity = new LinkedHashMap<>();
        entity.put("type", "breakpoint_rule");
        entity.put("id", rule.get("breakpoint_rule_id"));
        entity.put("label", rule.get("label"));
        entity.put("status", rule.get("status"));
        return entity;
    }

    private Map<String, Object> error(String status, String ruleId, String message) {
        return Map.of(
                "ok", false,
                "status", status,
                "breakpoint_rule_id", ruleId,
                "message", message,
                "entities", List.of());
    }

    private String label(String objectName, String cmdName, String displayName) {
        return displayName.isBlank() ? objectName + "." + cmdName : displayName;
    }

    private boolean bool(Object value) {
        if (value instanceof Boolean bool) {
            return bool;
        }
        if (value instanceof Number number) {
            return number.intValue() != 0;
        }
        return Boolean.parseBoolean(str(value)) || "1".equals(str(value));
    }

    private Object firstNonNull(Object... values) {
        for (Object value : values) {
            if (value != null) {
                return value;
            }
        }
        return null;
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value);
    }
}

File: python-debugger/app/services/agent_breakpoint_service.py
from app.db.database import get_db
from app.services.debug_service import declare_breakpoint_rule, list_breakpoints, normalize, row_to_dict
from app.utils.json_utils import loads
from app.utils.time_utils import now_iso


def declare_breakpoint(payload):
    return declare_breakpoint_rule(payload)


def list_breakpoint_rules(session_id=None):
    rules = [agent_rule(row) for row in list_breakpoints(session_id)]
    return {
        "ok": True,
        "breakpoint_rules": rules,
        "entities": [entity(rule) for rule in rules],
    }


def get_breakpoint_rule(rule_id):
    row = breakpoint_row(rule_id)
    if not row:
        return error("not_found", rule_id, "断点规则不存在。")
    return response(agent_rule(row), "断点规则已读取。")


def set_breakpoint_rule_enabled(rule_id, enabled):
    if not breakpoint_row(rule_id):
        return error("not_found", rule_id, "断点规则不存在。")
    db = get_db()
    db.execute(
        "UPDATE breakpoint SET enabled=?, updated_at=? WHERE id=?",
        (1 if enabled else 0, now_iso(), rule_id),
    )
    db.commit()
    return response(agent_rule(breakpoint_row(rule_id)), "断点规则已启用。" if enabled else "断点规则已禁用。")


def delete_breakpoint_rule(rule_id):
    row = breakpoint_row(rule_id)
    if not row:
        return error("not_found", rule_id, "断点规则不存在。")
    rule = agent_rule(row)
    get_db().execute("DELETE FROM breakpoint WHERE id=?", (rule_id,))
    get_db().commit()
    rule["status"] = "cancelled"
    return response(rule, "断点规则已取消。")


def breakpoint_row(rule_id):
    row = get_db().execute("SELECT * FROM breakpoint WHERE id=?", (rule_id,)).fetchone()
    if not row:
        return None
    return normalize(row_to_dict(row))


def agent_rule(row):
    object_name = row.get("objectName") or row.get("object_name") or ""
    cmd_name = row.get("cmdName") or row.get("cmd_name") or ""
    display_name = row.get("displayName") or row.get("display_name") or ""
    match_mode = row.get("matchMode") or row.get("match_mode") or "command_only"
    match = {
        "type": "parameters" if match_mode == "params_condition" else "interface",
        "mode": match_mode,
    }
    if match_mode == "params_condition":
        match["conditions"] = row.get("conditions") or loads(row.get("conditions_json"), [])
    return {
        "breakpoint_rule_id": row.get("id"),
        "label": display_name or f"{object_name}.{cmd_name}",
        "status": "armed" if is_enabled(row.get("enabled")) else "disabled",
        "target": {
            "object": object_name,
            "command": cmd_name,
            "display_name": display_name,
            "session_id": row.get("sessionId") or row.get("session_id"),
        },
        "match": match,
        "hit_count": row.get("hitCount", row.get("hit_count", 0)),
        "source_type": row.get("sourceType") or row.get("source_type"),
        "created_at": row.get("createdAt") or row.get("created_at"),
        "updated_at": row.get("updatedAt") or row.get("updated_at"),
    }


def response(rule, message):
    result = dict(rule)
    result["ok"] = True
    result["message"] = message
    result["entities"] = [entity(rule)]
    return result


def entity(rule):
    return {
        "type": "breakpoint_rule",
        "id": rule["breakpoint_rule_id"],
        "label": rule["label"],
        "status": rule["status"],
    }


def error(status, rule_id, message):
    return {
        "ok": False,
        "status": status,
        "breakpoint_rule_id": rule_id,
        "message": message,
        "entities": [],
    }


def is_enabled(value):
    if isinstance(value, str):
        return value.lower() not in ("", "0", "false")
    return bool(value)
